src/events/etcd.py

```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Limited
+# See LICENSE file for licensing details.
+
+"""Etcd related and core event handlers."""
+
+import logging
+from typing import TYPE_CHECKING
+
+import ops
+from ops import Object
+from ops.charm import (
+    LeaderElectedEvent,
+    RelationChangedEvent,
+    RelationCreatedEvent,
+    RelationDepartedEvent,
+    RelationJoinedEvent,
+)
+
+from literals import PEER_RELATION, Status
+from managers.cluster import ClusterManager
+from managers.config import ConfigManager
+
+if TYPE_CHECKING:
+    from charm import EtcdOperatorCharm
+
+logger = logging.getLogger(__name__)
+
+
+class EtcdEvents(Object):
+    """Handle all base and etcd related events."""
+
+    def __init__(self, charm: "EtcdOperatorCharm"):
+        super().__init__(charm, key="etcd_events")
+        self.charm = charm
+
+        # --- MANAGERS ---
+        self.cluster_manager = ClusterManager()
+        self.config_manager = ConfigManager(
+            state=self.charm.state, workload=self.charm.workload, config=self.charm.config
+        )
+
+        # --- Core etcd charm events ---
+
+        self.framework.observe(self.charm.on.install, self._on_install)
+        self.framework.observe(self.charm.on.start, self._on_start)
+        self.framework.observe(self.charm.on.config_changed, self._on_config_changed)
+        self.framework.observe(
+            self.charm.on[PEER_RELATION].relation_created, self._on_cluster_relation_created
+        )
+        self.framework.observe(
+            self.charm.on[PEER_RELATION].relation_joined, self._on_cluster_relation_joined
+        )
+        self.framework.observe(
+            self.charm.on[PEER_RELATION].relation_changed, self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            self.charm.on[PEER_RELATION].relation_departed, self._on_cluster_relation_departed
+        )
+        self.framework.observe(self.charm.on.leader_elected, self._on_leader_elected)
+        self.framework.observe(self.charm.on.update_status, self._on_update_status)
+
+    def _on_install(self, event: ops.InstallEvent) -> None:
+        """Handle install event."""
+        install = self.charm.workload.install()
+        if not install:
+            self.charm.set_status(Status.SERVICE_NOT_INSTALLED)
+            event.defer()
+            return
+
+    def _on_start(self, event: ops.StartEvent) -> None:
+        """Handle start event."""
+        # Make sure all planned units have joined the peer relation before starting the cluster
```

Comment:
Required to be able to assemble all the ip's for the cluster-endpoints in the `initial-cluster` config. If those can't be set, the unit can't join the cluster and could only be started as individual instance.

At a later point, this may be subject of change, as the cluster initialization needs to be robust for HA and failover. For now, this is just to be able to form an etcd-cluster on startup.

Reply:
This can be the case for the time being - while we haven't tackled scale-up, however we will need to abandon it and move to a rolling-(re)start approach when we start HA, where we should not be constrained with all planned units joining to start a cluster or** to add new units (think of a hardware failure in a new unit for ex). 

In which case this condition will no longer be relevant, as we should be able to bootstrap a cluster at first with only the first node and have additional nodes join the existing cluster using `--initial-cluster-state=existing` (as opposed to `=new`).

Reply:
Yes, that's exactly the plan.